text/0000-deref-move.md

```diff
@@ -0,0 +1,322 @@
+dereferencing were replaced with `DerefMove`, as `DerefMut` requires a mutable
+variable.
+
+This is an edge case, as it requires `Box<&mut T>`, which is an unlikely type,
+and in most cases it can be trivially fixed by making the variable mutable. It's
+possible there are more complex cases (such as if the `Box` itself is stored in
+another smart pointer) where this is not easily fixed, but I would be
+unsurprised if there were few, if any, instances of this in extant code.
+Moreover, the behaviour is arguably a bug because it is taking a mutable borrow
+of an immutable variable (and one can modify the above example to get an error
+message implying exactly that), though the semantics are more like the unique
+immutable borrows used by closures.
+
+As a result, I propose that this breaking change be accepted after verifying its
+effect on Cargo packages.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+## Copy handling
+
+The handling of copies is the real complexity of `DerefMove`. Currently, one can
+copy out of a `Box` without consuming it, so this desirable behaviour needs to
+be preserved. The proposal is to do so by special casing in the compiler.
+However, it could also be done via, for instance, a separate method in the Deref
+trait definition:
+
+```rust
+   fn deref_copy(&self) -> Self::Target where Self::Target : Copy {
+       return self.deref();
+   }
+```
+
+This still requires special-casing in the compiler, but it is less magical. One
+could even envision a `DerefCopy` trait with similar purpose.
+
+These approaches would, however, make diverging implementations (that is,
+implementations for which the behaviour of the various deref traits is
+unintuitive because they don't actually refer to the same underlying value) more
+difficult to write, and we want to avoid that. And they would not fundamentally
+solve the special casing in the compiler; to do that, we would really need a
+more advanced type system allowing us to condition the type of `self` in
+`deref_move` on whether or not `Target : Copy`.
+
+With negative trait bounds, we could also bound `DerefMove` on `Target : !Copy`,
+eliminating the problem, at the cost of requiring every implementation to also
+bound on `!Copy`.
+
+## Status quo
+
+The status quo is disfavoured because it requires special-casing `Box` in the
+compiler, privileging it over other smart pointers. While there's no obvious
+call in the stdlib for `DerefMut` other than `Box` (see below), it prevents
+other library authors from writing similar code.
+
+## `IndexMove` trait
+
+One option considered in the early discussion was an `IndexMove` trait, similar
+to `DerefMove`, which would consume a collection and return a single element. I
+have not included this in this RFC because no compelling use case was presented,
+and the behaviour of `let s = v[0];` actually consuming `v` seems likely to add
+cognitive overhead to the language.
```

Comment:
Oh i just thought of a use-case; arrays. IndexMove is necessary to get e.g. a Box out of an Array. But in that case into_iter or slice patterns are probably better.

Reply:
Well, there's definitely a use case for an Index trait which *returns* things by value.  For example, consider a type that wants to expose an API similar to `Vec<bool>`, but is implemented as a bit set, so indexing into it can't return an existing `&bool` reference.  However, you wouldn't usually want to have `self` be by value – but afaik that provides the most flexibility, since you could always do `impl<'a> DerefMove for &'a MyBitSet`.  (By itself, this would only allow reading values, not writing them, but #997 also proposed an `IndexSet` that would support writing.)

Reply:
IndexGet is the intended solution to that problem, but yes it could just be provided through IndexMove.

Reply:
I initially worked on all versions proposed by #997, but quickly abandoned that when I realize that the `IndexGet` and `IndexSet` make more sense as a separate RFC. Note that you couldn't impl `DerefMove` on a reference since it would have to have the same `Target` as the provided `Deref` instance; I don't think that the same restriction would necessarily apply to `IndexMove`, however.